c/parallel/src/jit_templates/templates/operation.h

```diff
@@ -13,20 +13,60 @@
 #ifndef _CCCL_C_PARALLEL_JIT_TEMPLATES_PREPROCESS
 #  include <cuda/std/cstddef>
 #  include <cuda/std/type_traits>
+#  include <cuda/std/utility>
 
 #  include <cccl/c/types.h>
 #endif
 
 #include "../mappings/operation.h"
 #include "../mappings/type_info.h"
 
+// Helper to define the C function pointer type: void (*)(void*..., void*)
+template <typename Seq>
+struct MakeVoidPtrFuncPtr;
+template <cuda::std::size_t... Is>
+struct MakeVoidPtrFuncPtr<cuda::std::index_sequence<Is...>>
+{
+  // Helper type that is always void*
+  template <cuda::std::size_t>
+  using void_ptr_t = const void*;
+  // Define the function pointer type
+  using type = void (*)(void_ptr_t<Is>..., void*); // Last void* is for result
+};
+
+// Helper to define the C function pointer type for stateful ops: void (*)(void* state, void*..., void*)
+template <typename Seq>
+struct MakeVoidPtrStatefulFuncPtr;
+template <cuda::std::size_t... Is>
+struct MakeVoidPtrStatefulFuncPtr<cuda::std::index_sequence<Is...>>
+{
+  template <cuda::std::size_t>
+  using void_ptr_t = const void*;
+  using type       = void (*)(void*, void_ptr_t<Is>..., void*); // First void* is state, last is result
+};
+
 template <typename Tag, cccl_op_t_mapping Operation, cccl_type_info_mapping RetT, cccl_type_info_mapping... ArgTs>
 struct stateless_user_operation
 {
-  __device__ decltype(RetT)::Type operator()(decltype(ArgTs)::Type... args) const
+  // Note: The user provided C f  unction (Operation.operation) must match the signature:
+  // void (void* arg1, ..., void* argN, void* result_ptr)
+  __device__ typename decltype(RetT)::Type operator()(typename decltype(ArgTs)::Type... args) const
   {
-    return reinterpret_cast<decltype(RetT)::Type (*)(decltype(ArgTs)::Type...)>(Operation.operation)(
-      std::move(args)...);
+    using Indices        = cuda::std::make_index_sequence<sizeof...(ArgTs)>;
+    using TargetCFuncPtr = typename MakeVoidPtrFuncPtr<Indices>::type;
+
+    // Cast the stored operation pointer (assumed to be void* or compatible)
+    auto c_func_ptr = reinterpret_cast<TargetCFuncPtr>(Operation.operation);
+
+    // Prepare storage for the result
+    typename decltype(RetT)::Type result;
+    // Get a void pointer to the result storage
+    void* result_ptr = static_cast<void*>(&result);
+
+    // Call the C function, casting argument addresses to void*
+    c_func_ptr((static_cast<void*>(const_cast<cuda::std::remove_reference_t<decltype(args)>*>(&args)))..., result_ptr);
```

Comment:
`const_cast<void*>(static_cast<const void*>(&args))...` should do the same, no?

Reply:
Yes - fixed